### PR TITLE
Test/improve code coverage

### PR DESCRIPTION
**Pull request by Saadha123**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

## Summary by CodeRabbit

* **Tests**
  * Massive expansion of automated tests across the controller: handlers, validators, storage (memory/SQLite), control plane/client, certs, API keys, policy/XDS, metrics, logging, LLM utilities, deployment/MCP, utilities and many edge cases — improving reliability, concurrency and error-path validation.
* **Chores**
  * Robustified client shutdown/state transitions to avoid locking/race issues.

### CODE REVIEW

File: gateway/gateway-controller/pkg/config/api_validator_test.go [hunk 1]
```diff
@@ -0,0 +1,703 @@
...
+		wantError bool
+		errField  string
+	}{
+		{name: "Valid URL", mainURL: stringPtr("http://backend:8080"), mainRef: nil, wantError: false},
+		{name: "Valid HTTPS URL", mainURL: stringPtr("https://backend:8443"), mainRef: nil, wantError: false},
+		{name: "Both URL and Ref set", mainURL: stringPtr("http://x"), mainRef: stringPtr("ref"), wantError: true, errField: "spec.upstream.main"},
+		{name: "Empty URL", mainURL: stringPtr(""), mainRef: nil, wantError: true, errField: "spec.upstream.main.url"},
+		{name: "Invalid URL scheme", mainURL: stringPtr("ftp://x"), mainRef: nil, wantError: true, errField: "spec.upstream.main.url"},
+		{name: "URL without host", mainURL: stringPtr("http:///path"), mainRef: nil, wantError: true, errField: "spec.upstream.main.url"},
+	}
```
[the remark this answers is not in the document]
Comment (Saadha123) [reply gateway/gateway-controller/pkg/config/api_validator_test.go]: Fixed in commit 600621ce80842672868fd11090c7295d708f0d7d

File: gateway/gateway-controller/pkg/controlplane/client_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+
+	// Close the connection directly (avoid Close() which has a potential deadlock)
+	// This tests the connection cleanup path
+	client.state.mu.Lock()
+	if client.state.Conn != nil {
+		client.state.Conn.Close()
+		client.state.Conn = nil
+		client.state.Current = Disconnected
+	}
+	client.state.mu.Unlock()
```
[the remark this answers is not in the document]
Comment (Saadha123) [reply gateway/gateway-controller/pkg/controlplane/client_integration_test.go]: Fixed in commit 3e9cecf97217760c0631c4c9e4434883928e72d9

File: gateway/gateway-controller/pkg/utils/api_utils_test.go [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+		defer server.Close()
+
+		cfg := PlatformAPIConfig{
+			BaseURL: server.URL,
+			Token:   "test-token",
+		}
+		svc := NewAPIUtilsService(cfg, logger)
+		// Verify service was created - actual call requires models.StoredConfig
+		assert.NotNil(t, svc)
+	})
```
[the remark this answers is not in the document]
Comment (Saadha123) [reply gateway/gateway-controller/pkg/utils/api_utils_test.go]: Fixed in commit c9c0fdd7d422f61909b658df8a9eaadac79f7e17
